### PR TITLE
[CodeCamp2023-367] Add pp_mobileseg model

### PR DESCRIPTION
**Pull request by Yang-Changhui**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMDet3D.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @Yang-Changhui,

We'd like to express our appreciation for your valuable contributions to the mmsegmentation. Your efforts have significantly aided in enhancing the project's quality.
It is our pleasure to invite you to join our community thorugh Discord_Special Interest Group (SIG) channel. This is a great place to share your experiences, discuss ideas, and connect with other like-minded people. To become a part of the SIG channel, send a message to the moderator, OpenMMLab, briefly introduce yourself and mention your open-source contributions in the #introductions channel. Our team will gladly facilitate your entry. We eagerly await your presence. Please follow this link to join us: ​https://discord.gg/UjgXkPWNqA.

If you're on WeChat, we'd also love for you to join our community there. Just add our assistant using the WeChat ID: openmmlabwx. When sending the friend request, remember to include the remark "mmsig + Github ID".

Thanks again for your awesome contribution, and we're excited to have you as part of our community!

File: mmseg/models/backbones/strideformer.py [hunk 1]
```diff
@@ -0,0 +1,765 @@
...
+from mmcv.cnn.bricks.transformer import build_dropout
+from mmseg.registry import MODELS
+
+
+class StrideFormer(BaseModule):
+    def __init__(self,
+                 cfg1,
+                 cfg2,
+                 cfg3,
+                 cfg4,
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:20]: Generally, we should use meaningful parameter names.

File: mmseg/models/backbones/strideformer.py [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+from mmengine.logging import print_log
+from mmengine.model import BaseModule
+from mmengine.runner.checkpoint import CheckpointLoader, load_state_dict
+from mmcv.cnn import build_conv_layer, build_norm_layer, build_activation_layer, ConvModule
+from mmcv.cnn.bricks.transformer import build_dropout
+from mmseg.registry import MODELS
+
+
+class StrideFormer(BaseModule):
+    def __init__(self,
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:16]: We should add docstring.

File: mmseg/models/backbones/strideformer.py [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+            bias=None if bias_attr else False)
+        self.act = build_activation_layer(act_cfg) if act_cfg is not None else nn.Identity()
+        self.bn = build_norm_layer(norm_cfg, out_channels)[1] \
+            if norm_cfg is not None else nn.Identity()
+
+    def forward(self, x):
+        x = self.conv(x)
+        x = self.bn(x)
+        x = self.act(x)
+        return x
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:259]: Its function is the same as `ConvModule`, might use `ConvModule` instead.

File: mmseg/models/backbones/strideformer.py [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+                 dilation=1,
+                 groups=1,
+                 ):
+        super().__init__()
+        self.conv_norm = ConvModule(in_channels, out_channels, ks, stride, pad, dilation, groups, False,
+                                    norm_cfg=dict(type='BN'), act_cfg=None)
+
+    def forward(self, inputs):
+        out = self.conv_norm(inputs)
+        return out
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:278]: It's no need to define this class.

File: mmseg/models/backbones/strideformer.py [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+        self.conv_norm = ConvModule(in_c, out_c, filter_size, stride, padding, dilation, num_groups, False,
+                                    norm_cfg=dict(type='BN'), act_cfg=None)
+        self.if_act = if_act
+        self.act = _create_act(act)
+
+    def forward(self, x):
+        x = self.conv_norm(x)
+        if self.if_act:
+            x = self.act(x)
+        return x
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:302]: Its function is the same as `ConvModule`, might use `ConvModule` instead.

File: mmseg/models/backbones/strideformer.py [hunk 6]
```diff
@@ -0,0 +1,765 @@
...
+
+
+class ResidualUnit(nn.Module):
+    def __init__(self,
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:306]: Might add a docstring.

File: mmseg/models/backbones/strideformer.py [hunk 7]
```diff
@@ -0,0 +1,765 @@
...
+            padding=0)
+
+    def forward(self, x):
+        identity = x
+        x = self.avg_pool(x)
+        x = self.conv1(x)
+        x = self.relu(x)
+        x = self.conv2(x)
+        # Hardsigmoid
+        x = x.div(5).add(0.5).clamp(0, 1)
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:390]: The parameters of this operation should be configurable.

File: mmseg/models/backbones/strideformer.py [hunk 8]
```diff
@@ -0,0 +1,765 @@
...
+            stride=1,
+            padding=0)
+        self.relu = build_activation_layer(act_cfg)
+        self.conv2 = build_conv_layer(
+            conv_cfg,
+            in_channels=channel // reduction,
+            out_channels=channel,
+            kernel_size=1,
+            stride=1,
+            padding=0)
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:381]: Might instead with two `ConvModule`.

File: mmseg/models/backbones/strideformer.py [hunk 9]
```diff
@@ -0,0 +1,765 @@
...
+            self.pos_embed,
+            size=(N,),
+            mode='linear',
+            align_corners=False
+        )
+        return x
+
+
+class Sea_Attention(nn.Module):
+    def __init__(self,
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:486]: Might rename to `SeaAttention`.

File: mmseg/models/backbones/strideformer.py [hunk 10]
```diff
@@ -0,0 +1,765 @@
...
+        x = self.dwconv(x)
+        x = self.act(x)
+        x = self.drop(x)
+        x = self.fc2(x)
+        x = self.drop(x)
+        return x
+
+
+class Fusion_block(nn.Module):
+    def __init__(self, inp, oup, embed_dim, activations=None) -> None:
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:623]: Might rename to `FusionBlock`.

Comment (xiexinch) [reply mmseg/models/backbones/strideformer.py:623]: And the names of parameters are not readable, such as `inp`, `oup`. It suggests changing to names more readable.

File: mmseg/models/backbones/strideformer.py [hunk 11]
```diff
@@ -0,0 +1,765 @@
...
+def _create_act(act):
+    if act == "hardswish":
+        return nn.Hardswish()
+    elif act == "relu":
+        return nn.ReLU()
+    elif act is None:
+        return None
+    else:
+        raise RuntimeError(
+            "The activation function is not supported: {}".format(act))
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:753]: Can we replace it with `build_activation_layer`?

File: mmseg/models/backbones/strideformer.py [hunk 12]
```diff
@@ -0,0 +1,765 @@
...
+@MODELS.register_module()
+class MobileSeg_Base(StrideFormer):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+
+@MODELS.register_module()
+class MobileSeg_Tiny(StrideFormer):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:765]: We might add `StrideFormer` to the `MODEL` register and delete these two classes.

File: mmseg/models/backbones/strideformer.py (hunk 11 above)
Comment (Yang-Changhui) [reply mmseg/models/backbones/strideformer.py:753]: build_activation_layer好像没有nn.Hardswish()有这个类

Comment (xiexinch) [reply mmseg/models/backbones/strideformer.py:753]: 那可以单独写一个包装类，然后注册进去吗，这样写法就比较通用了

File: mmseg/models/backbones/strideformer.py [hunk 13]
```diff
@@ -0,0 +1,753 @@
...
+                 mlp_ratios=[2, 4],
+                 drop_path_rate=0.1,
+                 act_cfg=dict(type='ReLU'),
+                 inj_type='AAM',
+                 out_channels=256,
+                 dims=(128, 160),
+                 out_feat_chs=None,
+                 stride_attention=True,
+                 pretrained=None,
+                 init_cfg=None
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:33]: Might add type hints

File: mmseg/models/backbones/strideformer.py (hunk 6 above)
Comment (xiexinch) [reply mmseg/models/backbones/strideformer.py:236]: And type hints are needed.

File: mmseg/models/backbones/strideformer.py [hunk 14]
```diff
@@ -0,0 +1,753 @@
...
+    def forward(self, x1):
+        x1 = x1 + self.drop_path(self.attn(x1))
+        x1 = x1 + self.drop_path(self.mlp(x1))
+
+        return x1
+
+
+class SqueezeAxialPositionalEmbedding(nn.Module):
+    def __init__(self, dim, shape):
+        super().__init__()
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:399]: Might add a docstring.

File: mmseg/models/backbones/strideformer.py [hunk 15]
```diff
@@ -0,0 +1,753 @@
...
+
+class SeaAttention(nn.Module):
+    def __init__(self,
+                 dim,
+                 key_dim,
+                 num_heads,
+                 attn_ratio=4.,
+                 act_cfg=None,
+                 norm_cfg=dict(type='BN'),
+                 stride_attention=False):
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:422]: Might add docstring and type hints.

File: mmseg/models/backbones/strideformer.py [hunk 16]
```diff
@@ -0,0 +1,753 @@
...
+class HSigmoid(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.relu = nn.ReLU6()
+
+    def forward(self, x):
+        return self.relu(x + 3) / 6
+
+
+MODELS.register_module(module=HSigmoid, name='HSigmoid')
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:728]:
```suggestion
@MODELS.register_module()
class HSigmoid(nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = nn.ReLU6()

    def forward(self, x):
        return self.relu(x + 3) / 6
```

File: mmseg/models/backbones/strideformer.py [hunk 17]
```diff
@@ -0,0 +1,753 @@
...
+class hardswish(nn.Module):
+    def __init__(self, inplace=False):
+        super().__init__()
+        self.relu = nn.Hardswish(inplace=inplace)
+
+    def forward(self, x):
+        return self.relu(x)
+
+
+MODELS.register_module(module=hardswish, name='hardswish')
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:740]:
```suggestion
@MODELS.register_module()
class Hardswish(nn.Module):
    def __init__(self, inplace=False):
        super().__init__()
        self.relu = nn.Hardswish(inplace=inplace)

    def forward(self, x):
        return self.relu(x)

```

Comment (xiexinch) [reply mmseg/models/backbones/strideformer.py:740]: Might add a docstring to explain that it's a wrapper to `torch.nn.Hardswish`.

File: mmseg/models/backbones/strideformer.py [hunk 18]
```diff
@@ -0,0 +1,753 @@
...
+    def __init__(self, slope=0.2, offset=0.5, inplace=False):
+        super().__init__()
+        self.slope = slope
+        self.offset = offset
+
+    def forward(self, x):
+        return x.mul(self.slope).add(self.offset).clamp(0, 1)
+
+
+MODELS.register_module(module=Hardsigmoid, name='Hardsigmoid')
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:753]:
```suggestion
@MODELS.register_module()
class Hardsigmoid(nn.Module):
    def __init__(self, slope=0.2, offset=0.5, inplace=False):
        super().__init__()
        self.slope = slope
        self.offset = offset

    def forward(self, x):
        return x.mul(self.slope).add(self.offset).clamp(0, 1)
```

Comment (xiexinch) [reply mmseg/models/backbones/strideformer.py:753]: Also, we should add a docstring.

File: mmseg/models/decode_heads/pp_mobileseg_head.py [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+                 num_classes,
+                 in_channels,
+                 use_dw=True,
+                 dropout_ratio=0.1,
+                 align_corners=False,
+                 upsample='intepolate',
+                 out_channels=None,
+                 conv_cfg=dict(type='Conv'),
+                 act_cfg=dict(type='ReLU'),
+                 norm_cfg=dict(type='BN')
```
Comment (xiexinch) [inline mmseg/models/decode_heads/pp_mobileseg_head.py:22]: We should add a docstring and type hints.

File: mmseg/models/backbones/strideformer.py [hunk 20]
```diff
@@ -0,0 +1,852 @@
...
+            depths(list, optional): describes the depth of the attention block.
+                i,e: M,N.
+            num_heads(int, optional): The number of heads of the attention
+                blocks.
+            attn_ratios(int, optional): The expand ratio of V.
+            mlp_ratios(list, optional): The ratio of mlp blocks.
+            drop_path_rate(float, optional): The drop path rate in attention
+                block.
+            act_cfg(nn.Layer, optional): The activation layer of AAM.
+            inj_type(string, optional): The type of injection/AAM.
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:60]: Might add a brief introduction to 'AAM'.

File: mmseg/models/backbones/strideformer.py [hunk 21]
```diff
@@ -0,0 +1,852 @@
...
+@MODELS.register_module()
+class HSigmoid(nn.Module):
+
+    def __init__(self):
+        super().__init__()
+        self.relu = nn.ReLU6()
+
+    def forward(self, x):
+        return self.relu(x + 3) / 6
+
```
Comment (xiexinch) [inline mmseg/models/backbones/strideformer.py:830]: Might use the implementation in mmcv https://github.com/open-mmlab/mmcv/blob/main/mmcv/cnn/bricks/hsigmoid.py#L10

**Review by xiexinch — APPROVED** [review]

**Pull request merged**